tirocks/src/table_properties/builtin.rs

```diff
@@ -0,0 +1,428 @@
+            );
+            str::from_utf8(s(buf))
+        }
+    }
+
+    /// The name of the merge operator used in this table.
+    /// If no merge operator is used, `merge_operator_name` will be "nullptr".
+    #[inline]
+    pub fn merge_operator_name(&self) -> std::result::Result<&str, Utf8Error> {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_table_properties_get_merge_operator_name(
+                self as *const _ as _,
+                &mut buf,
+            );
+            str::from_utf8(s(buf))
+        }
+    }
+
+    /// The name of the prefix extractor used in this table
+    /// If no prefix extractor is used, `prefix_extractor_name` will be "nullptr".
+    #[inline]
+    pub fn prefix_extractor_name(&self) -> std::result::Result<&str, Utf8Error> {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_table_properties_get_prefix_extractor_name(
+                self as *const _ as _,
+                &mut buf,
+            );
+            str::from_utf8(s(buf))
+        }
+    }
+
+    /// The names of the property collectors factories used in this table
+    /// separated by commas
+    /// {collector_name[1]},{collector_name[2]},{collector_name[3]} ..
+    #[inline]
+    pub fn property_collectors_names(&self) -> std::result::Result<&str, Utf8Error> {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_table_properties_get_property_collectors_names(
+                self as *const _ as _,
+                &mut buf,
+            );
+            str::from_utf8(s(buf))
+        }
+    }
+
+    /// The compression algo used to compress the SST files.
+    #[inline]
+    pub fn compression_name(&self) -> std::result::Result<&str, Utf8Error> {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_table_properties_get_compression_name(
+                self as *const _ as _,
+                &mut buf,
+            );
+            str::from_utf8(s(buf))
+        }
+    }
+
+    /// Compression options used to compress the SST files.
+    #[inline]
+    pub fn compression_options(&self) -> std::result::Result<&str, Utf8Error> {
+        unsafe {
+            let mut buf = r(&[]);
+            tirocks_sys::crocksdb_table_properties_get_compression_options(
+                self as *const _ as _,
+                &mut buf,
+            );
+            str::from_utf8(s(buf))
+        }
+    }
+}
+
+impl Debug for TableProperties {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let mut len = 0;
+        unsafe {
+            let s =
+                tirocks_sys::crocksdb_table_properties_to_string(self as *const _ as _, &mut len);
+            let buf = slice::from_raw_parts(s as *mut u8, len);
+            let res = write!(f, "{}", String::from_utf8_lossy(buf));
+            libc::free(s as _);
+            res
+        }
+    }
+}
+
+enum Collection<'a> {
```

Comment:
use `std::borrow::Cow`?

Reply:
`Cow` is not useful in this case. First, to form a `Cow` relation, there needs to be two types, it will goes back to the days that uses two types to represent the same thing. Second, there is no such requirement to convert the collection from borrowed to owned.